[SYCL][Graph] Expand on design doc (#224)
reble/llvm: sycl/doc/design/CommandGraph.md

# Command-Graph Extension

This document describes the implementation design of the

> [SYCL Graph Extension](https://github.com/intel/llvm/pull/5626).

Reviewer: A direct link to the file could be probably clearer that a link to the PR.

Reviewer 2: Makes sense, we can update this once the PR is merged.

Author: Agree that once the PR is merged we should update this to the direct link

command-list multiple times.

### UR Command-Buffer Experimental Feature

The command-buffer concept has been introduced to UR as an
[experimental feature](https://oneapi-src.github.io/unified-runtime/core/api.html#command-buffer-experimental)
with the following entry-points:

| Function                                | Description |
| --------------------------------------- | ----------- |
| `urCommandBufferCreateExp`              | Create a command-buffer. |
| `urCommandBufferRetainExp`              | Incrementing reference count of command-buffer. |
| `urCommandBufferReleaseExp`             | Decrementing reference count of command-buffer. |
| `urCommandBufferFinalizeExp`            | No more commands can be appended, makes command-buffer ready to enqueue on a command-queue. |
| `urCommandBufferAppendKernelLaunchExp`  | Append a kernel execution command to command-buffer. |
| `urCommandBufferAppendMemcpyUSMExp`     | Append a USM memcpy command to the command-buffer. |
| `urCommandBufferAppendMembufferCopyExp` | Append a mem buffer copy command to the command-buffer. |
| `urCommandBufferEnqueueExp`             | Submit command-buffer to a command-queue for execution. |

> See the [UR EXP-COMMAND-BUFFER](https://oneapi-src.github.io/unified-runtime/core/EXP-COMMAND-BUFFER.html)

Author: Need to double check this link when https://github.com/oneapi-src/unified-runtime/pull/591/ is merged, won't be valid until then

Author: Live now and works 😃

specification for more details.

## Design

![Basic architecture diagram.](images/SYCL-Graph-Architecture.svg)

There are two sets of user facing interfaces that can be used to create a
command-graph object: Explicit and Record & Replay API. Within the runtime they
share a common infrastructure.

## Nodes & Edges

Nodes are represented by an internal `detail::node_impl` class which
stores a `sycl::detail::CG` object for the command-group that the node
represents. When a user adds a node to a graph using the explicit
`command_graph<modifiable>::add()` API passing a CGF, in the implementation a
`sycl::handler` object is created which finalizes the CGF, and after

> finalization the CG object from the handler is moved to the node.

Reviewer: This sentence is difficult to understand. 
What CGF stand for?
Maybe some extra punctuation could help understanding. e.g. ... and after finalization, the CG ...

Author: I've expanded this section to give more context and hopefully make it clearer, if you want to take another look

Reviewer: Far much clearer. Thanks :+1:

> finalization the CG object from the handler is moved to the node.

Reviewer 3: I think it could be worth clarifying in this section that we're preventing the `CG` object from being submitted for execution as normal inside `handler::finalize()` and then storing it in the graph as a new node instead.

Author: Done, i've also expanded this section quite a bit

For creating a node in the graph using queue recording mode. When the

> `sycl::handler` from a queue submissions is finalized, if the queue the

Reviewer 3: ```suggestion
`sycl::handler` from a queue submission is finalized, if the queue the
```

handler was created from is in the recording mode, then the command-group from

> finalization is added to the graph associated with the queue as a new node.

Reviewer 4: These sentences are not clear to me.

Author: I've expanded this section to give more context and hopefully make it clearer, if you want to take another look

Edges are stored in each node as lists of predecessor and successor nodes.

> ## Sorting

Reviewer 4: Should we clarify what we are sorting? E.g. "Node Sorting" or "Execution Order".

Author: Good point, gone with "Execution Order"

The current way graph nodes are linearized into execution order is using a
reversed depth-first algorithm. Alternative algorithms, such as breadth-first,
are possible and may give better performance on certain workloads/hardware. We
are looking into giving the user control of this implementation detail.

## Scheduler Integration

When there are no requirements from accessors in a command-graph submission,
the scheduler is bypassed and the underlying UR command-buffer is directly
enqueued to a UR queue. If there are accessor requirements, the UR
command-buffer for the executable graph needs to be enqueued by the scheduler.

When individual graph nodes have requirements from SYCL accessors, the
underlying `sycl::detail::CG` object stored in the node is copied and passed to
the scheduler for adding to the UR command-buffer, otherwise the node can

> be appended directly as a command in the UR command-buffer.

Reviewer 3: Might be worth noting here that this is in keeping with the existing behaviour of the handler/normal queue submissions.

## Memory handling: Buffer and Accessor

There is no extra support for Graph-specific USM allocations in the current
proposal. Memory operations will be supported subsequently by the current
implementation starting with `memcpy`.

> Buffers and accessors are supported in a command-graph. Following restrictions

Reviewer 3: ```suggestion
Buffers and accessors are supported in a command-graph. The following restrictions
```

are required to adapt buffers and their lifetime to a lazy work execution model:

> - Lifetime of a buffer with host data will be extended by copying the underlying

Reviewer 3: ```suggestion
- The lifetime of a buffer with host data will be extended by copying the underlying
```

data.

> - Host accessor on buffer that are used by a command-graph are prohibited.

Reviewer 3: ```suggestion
- Host accessors on buffers that are currently used by a command-graph are prohibited.
```
I think the wording here was a bit ambiguous in that it could suggest that a buffer which had been used in a command-graph but that command-graph has since been destroyed would be prohibited, which I don't think we want.

- Copy-back behavior on destruction of a buffer is prohibited.

## Backend Implementation

Implementation of [UR command-buffers](#UR-command-buffer-experimental-feature)

> for each of the supported SYCL 2020 backends.

Reviewer: I'm not sure to understand this sentence.
In "Requirements" Section you mention that only Level Zero supports command-buffer, but here it seems that UR can enable other backends.

Author: It will be able to enable other backends in future, hopefully CUDA this year, so when can add new subsections here when they are done.

I've added an extra sentence below to make that clear though

Reviewer: Make sense.

* Prefix - Commands added to the start of the L0 command-list by L0 adapter.
* Suffix - Commands added to the end of the L0 command-list by L0 adapter.

These extra commands operate on L0 event synchronisation primitives, used by the
command-list to interact with the external UR wait-list and UR return event
required for the enqueue interface.

The `ur_exp_command_buffer_handle_t` class for this adapter contains a
*SignalEvent* which signals the completion of the command-list in the suffix,
and is reset in the prefix. This signal is detected by a new UR return event
created on UR command-buffer enqueue.

There is also a *WaitEvent* used by the `ur_exp_command_buffer_handle_t` class
in the prefix to wait on any dependencies passed in the enqueue wait-list.

```mermaid
flowchart TB
    subgraph Suffix
    id1[Signal the UR command-buffer SignalEvent]
    end
    subgraph User Added Commands
    id2[Command 1] -.-  id3[Command N]
    end
    subgraph Prefix
    id4[Reset SignalEvent] ~~~ id5[Barrier waiting on WaitEvent]
    end
```

For a call to `urCommandBufferEnqueueExp` with an `event_list` *EL*,
command-buffer *CB*, and return event *RE* our implementation has to create and

> submit two new command-lists for the above approach to work. One before

Reviewer 3: I don't know if this is too nitpicky but this might be misleading a reader to think that we actually create a new command list manually every time for these when we are getting them from the UR Queue which will likely just reuse one, and only create a new one in the worst case.

Author: Rephrased this to mention getting new command-lists from UR queue

the command-list with extra commands associated with *CB*, and the other
after *CB*.

```mermaid
flowchart TB
    subgraph L0 Command-list created on urCommandBufferEnqueueExp to execution before CB

>     id1[Barrier on EL than signals CB WaitEvent when completed]

Reviewer: ```suggestion
    id1[Barrier on EL that signals CB WaitEvent when completed]
```

# Command-Graph Extension

This document describes the implementation design of the
[SYCL Graph Extension](../extensions/proposed/sycl_ext_oneapi_graph.asciidoc).

A related presentation can be found
[here](https://www.youtube.com/watch?v=aOTAmyr04rM).

## Requirements

An efficient implementation of a lazy command-graph execution and its replay
requires extensions to the Unified Runtime (UR) layer. Such an extension is

> command-buffers, where a command-buffer object represents a series of operations

Reviewer 2: ```suggestion
UR command-buffer, where a command-buffer object represents a series of operations
```

Reviewer 3: I've removed the plural but didn't really like the way it was reading so I've modified the surrounding wording a little to read better.

to be enqueued to the backend device and their dependencies. A single
command-graph can be partitioned into more than one command-buffer by the runtime.

> We distinguish between backends that support command-buffer extensions and

Reviewer 2: ```suggestion
The SYCL Graph extension distinguishes between backends that support command-buffer extensions and
```

(CG) that is defined by a [command-group function](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#command-group-function-object)
(CGF).

Internally, a node is represented by the `detail::node_impl` class, and a command-group
by the `sycl::detail::CG` class. An instance of `detail::node_impl` stores a
`sycl::detail::CG` object for the command-group that the node represents.

A [command-group handler](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#handler)
lets the user define the operations that are to be performed in the command-group,
e.g. kernel execution, memory copy, host-task. In DPC++ an internal "finalization"
operation is done inside the `sycl::handler` implementation, which constructs a
CG object of a specific type. During normal operation, `handler::finalize()`
then passes the CG object to the scheduler, and a `sycl::event` object
representing the command-group is returned.

However during graph construction, inside `hander::finalize()` we prevent the
CG object from being submitted for execution as normal, and store CG in the

> graph as a new node instead.

Reviewer 2: ```suggestion
However during graph construction, inside `hander::finalize()` the
CG object is not submitted for execution as normal, but stored in the
graph as a new node instead.
```

When a user adds a node to a graph using the explicit
`command_graph<modifiable>::add()` API passing a CGF, in our graph runtime
implementation a `sycl::handler` object is constructed with a graph parameter
telling it to not submit the CG object to the scheduler on finalization.
This handler finalizes the CGF, and after finalization the CG object from the
handler is moved to the node.

For creating a node in the graph using queue recording mode. When the
`sycl::handler` from a queue submission is finalized, if the queue the
handler was created from is in the recording mode, then the handler knows
not to submit the CG object to the scheduler. Instead, the CG object is
added to the graph associated with the queue as a new node.

Edges are stored in each node as lists of predecessor and successor nodes.

## Execution Order

The current way graph nodes are linearized into execution order is using a
reversed depth-first sorting algorithm. Alternative algorithms, such as
breadth-first, are possible and may give better performance on certain
workloads/hardware. We are looking into giving the user control of this

> implementation detail.

Reviewer 2: ```suggestion
workloads/hardware. In the future there might be options for controlling this
implementation detail by the user.
```

Reviewer 3: Removed the `we` but reworked the wording slightly from the suggestion.

underlying `sycl::detail::CG` object stored in the node is copied and passed to
the scheduler for adding to the UR command-buffer, otherwise the node can
be appended directly as a command in the UR command-buffer. This is in-keeping
with the existing behaviour of the handler with normal queue submissions.

## Memory handling: Buffer and Accessor

There is no extra support for Graph-specific USM allocations in the current
proposal. Memory operations will be supported subsequently by the current
implementation starting with `memcpy`.

Buffers and accessors are supported in a command-graph. The following restrictions
are required to adapt buffers and their lifetime to a lazy work execution model:

- The lifetime of a buffer with host data will be extended by copying the underlying
data.
- Host accessors on buffers that are currently used by a command-graph are prohibited.
- Copy-back behavior on destruction of a buffer is prohibited.

## Backend Implementation

Implementation of [UR command-buffers](#UR-command-buffer-experimental-feature)
for each of the supported SYCL 2020 backends.

This is currently only Level Zero but more sub-sections will be added here as

> we implement other backends.

Reviewer 2: ```suggestion
other backends are implemented.
```

### Level Zero

The UR `urCommandBufferEnqueueExp` interface for submitting a command-buffer
takes a list of events to wait on, and returns an event representing the
completion of that specific submission of the command-buffer.

However, in the equivalent Level Zero function
[zeCommandQueueExecuteCommandLists](https://spec.oneapi.io/level-zero/latest/core/api.html#zecommandqueueexecutecommandlists)
there are no parameters to take a wait-list, and the only sync primitive
returned is blocking on host.

In order to get the UR command-buffer enqueue semantics we want with Level Zero,
the adapter implementation adds extra commands to the Level Zero command-list

> representing a UR command-buffer.

Reviewer 2: ```suggestion
In order to achieve the expected UR command-buffer enqueue semantics with Level Zero,
the adapter implementation adds extra commands to the Level Zero command-list
representing a UR command-buffer.
```

    end
```

For a call to `urCommandBufferEnqueueExp` with an `event_list` *EL*,
command-buffer *CB*, and return event *RE* our implementation has to submit two
new command-lists for the above approach to work. One before
the command-list with extra commands associated with *CB*, and the other
after *CB*. These two new command-lists are retrieved from the UR queue, which
will likely reuse existing command-lists and only create a new one in the worst
case.

```mermaid
flowchart TB
    subgraph L0 Command-list created on urCommandBufferEnqueueExp to execution before CB
    id1[Barrier on EL that signals CB WaitEvent when completed]
    end
    subgraph L0 Command-list created on urCommandBufferEnqueueExp to execution after CB
    id2[Barrier on CB SignalEvent that signals RE when completed]
    end
```

#### Drawbacks

There are two drawbacks of this approach to implementing UR command-buffers for
Level Zero:

1. We use 3x the command-list resources, if there are many UR command-buffers in
   flight, this may exhaust L0 driver resources. A trivial graph requires 3 L0
   command-lists and if we implement partitioning a graph into multiple UR

>    command-buffers, then each partition will be 3 L0 command-lists.

Reviewer 2: ```suggestion
1. 3x the command-list resources are used, if there are many UR command-buffers in
   flight, this may exhaust L0 driver resources. A trivial graph requires 3 L0
   command-lists and if we implement partitioning a graph into multiple UR
   command-buffers, then each partition will contain 3 L0 command-lists.
```

See the [UR EXP-COMMAND-BUFFER](https://oneapi-src.github.io/unified-runtime/core/EXP-COMMAND-BUFFER.html)
specification for more details.

## Design

![Basic architecture diagram.](images/SYCL-Graph-Architecture.svg)

There are two sets of user facing interfaces that can be used to create a
command-graph object: Explicit and Record & Replay API. Within the runtime they
share a common infrastructure.

## Nodes & Edges

A node in a graph is a SYCL [command-group](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#command-group)
(CG) that is defined by a [command-group function](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#command-group-function-object)
(CGF).

Internally, a node is represented by the `detail::node_impl` class, and a command-group
by the `sycl::detail::CG` class. An instance of `detail::node_impl` stores a
`sycl::detail::CG` object for the command-group that the node represents.

A [command-group handler](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#handler)
lets the user define the operations that are to be performed in the command-group,

> e.g kernel execution, memory copy, host-task. In DPC++ an internal "finalization"

Reviewer 2: ```suggestion
e.g. kernel execution, memory copy, host-task. In DPC++ an internal "finalization"
```

will likely reuse existing command-lists and only create a new one in the worst
case.

```mermaid
flowchart TB
    subgraph L0 Command-list created on urCommandBufferEnqueueExp to execution before CB
    id1[Barrier on EL that signals CB WaitEvent when completed]
    end
    subgraph L0 Command-list created on urCommandBufferEnqueueExp to execution after CB
    id2[Barrier on CB SignalEvent that signals RE when completed]
    end
```

#### Drawbacks

There are two drawbacks of this approach to implementing UR command-buffers for
Level Zero:

1. We use 3x the command-list resources, if there are many UR command-buffers in
   flight, this may exhaust L0 driver resources. A trivial graph requires 3 L0
   command-lists and if we implement partitioning a graph into multiple UR
   command-buffers, then each partition will be 3 L0 command-lists.

2. Each L0 command-list is submitted individually with a
   `ur_queue_handle_t_::executeCommandList` call which introduces serialization
   in the submission pipeline that is heavier than having a barrier or a
   `waitForEvents` on the same command-list. Resulting in additional latency when
   executing a UR command-buffer.

> We are consulting with L0 API owners about these problems and how the L0 API
> could be modified to achieve a 1:1 mapping of UR command-buffer to L0
> command-list.

Reviewer 2: ```suggestion
Future work will include exploring L0 API extensions to improve the mapping of 
UR command-buffer to L0 command-list.
```

Implementation of [UR command-buffers](#UR-command-buffer-experimental-feature)
for each of the supported SYCL 2020 backends.

This is currently only Level Zero but more sub-sections will be added here as
we implement other backends.

> ### Level Zero

Author: I should document these drawbacks to our implementation highlighted by the maintainer
https://github.com/intel/llvm/pull/9992#discussion_r1246262690
